Translate Unknown Prop Warning (#212)
reactjs/ru.reactjs.org: content/warnings/unknown-prop.md

---
title: "Предупреждение: неизвестный проп"
layout: single
permalink: warnings/unknown-prop.html
---

> Предупреждение о неизвестном пропе выдаётся, когда вы пытаетесь срендерить DOM-элемент с пропом, который React не может распознать как разрешённый DOM-атрибут или свойство. Вам следует убедиться, что DOM-элементы не получают по ошибке не относящиеся к ним пропсы.

Reviewer: Попробовал избежать _не_ два раза подряд
```suggestion
Предупреждение о неизвестном пропе выдаётся, когда вы пытаетесь срендерить DOM-элемент с пропом, который React не может распознать как разрешённый DOM-атрибут или свойство. Вам следует убедиться, что DOM-элементы не получают по ошибке пропсы, которые к ним не относятся.
```

> Предупреждение о неизвестном пропе выдаётся, когда вы пытаетесь срендерить DOM-элемент с пропом, который React не может распознать как разрешённый DOM-атрибут или свойство. Вам следует убедиться, что DOM-элементы не получают по ошибке пропсы, которые к ним не относятся.

Reviewer 2: Мне кажется, что вариант с переводом "отрендерить" менее режет слух. Как думаете?
```suggestion
Предупреждение о неизвестном пропе выдаётся, когда вы пытаетесь отрендерить DOM-элемент с пропом, который React не может распознать как разрешённый DOM-атрибут или свойство. Вам следует убедиться, что DOM-элементы не получают по ошибке пропсы, которые к ним не относятся.
```

3. React не распознаёт указанный атрибут. С большой вероятностью, это будет исправлено в будущих версиях React. На сегодняшний же день React удаляет неизвестные атрибуты, поэтому они не будут отрендерены.

4. Вы используете React-компонент, имя которого начинается со строчной, а не заглавной буквы. React интерпретирует его как DOM-тег, потому что [JSX-трасформация в React использует соглашение о регистре в наименованиях для отличия определённых пользователем компонентов от DOM-тегов](/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized).

---

Чтобы исправить ситуацию, составные компоненты должны получить все пропсы, которые предназначены только ему, а не его дочерним компонентам. Пример:

**Плохо:** Непредвиденный проп `layout` перенаправлен в тег `div`.

```js
function MyDiv(props) {
  if (props.layout === 'horizontal') {
    // Плохо! Потому что мы уверены, что "layout" не проп, с которым <div> работает.
    return <div {...props} style={getHorizontalStyle()} />
  } else {
    // Плохо! Потому что мы уверены, что "layout" не проп, с которым <div> работает.
    return <div {...props} style={getVerticalStyle()} />
  }
}
```

> **Хорошо:** Оператор расширения (`...`) может пригодиться, чтобы «вытянуть» отдельные переменные из пропсов, а оставшиеся пропсы поместить в переменную.

Reviewer: Хотя вытянуть звучит интересно

```suggestion
**Хорошо:** Оператор расширения (`...`) может пригодиться, чтобы получить отдельные переменные из пропсов, а оставшиеся пропсы поместить в переменную.
```

Author: @gcor я пытался использовать слово, лучше передающее _вытягивание_, а не простое _получение_ отдельных значений. В оригинале подчёркивается, что в новом объекте будут отсутствовать ключи, которым соответствуют явно заданные переменные.

Как считаешь, будет ли _**извлечь** переменные из пропсов_ звучать лучше? _Получить_ уж слишком далеко от исходного смысла.

Reviewer 3: Кстати да! _Излечь_ отлично подходит, правда не очень "звучит" с переменные (как можно извлечь переменную, может быть _в переменную_? Либо в кавычки завернуть.

Reviewer: Я бы ещё уточнил пару моментов.

1. Возможно стоит убрать «может пригодиться». Тут совет как делать хорошо, а не просто предложение.
2. Скорее стоит использовать _и_, как в оригинале, а не _а_, потому что тут скорее перечисление возможностей.

Оператор расширения (`...`) помогает извлечь часть пропсов и оставшиеся поместить в переменную.

На следующей строке стоит пример использования. Можно тогда прямо показать конкретно:
Оператор расширения (`...`) помогает извлечь часть пропсов (`layout`) и оставшиеся поместить в переменную (`rest`).

Author: @lex111 спасибо за отзыв про _извлечь_.

@gcor спасибо за советы.

С первым я согласен на :100: (я тут действительно утратил рекомендательный тон).

А вот со вторым -- нет. Я слышу, что в оригинале не перечисляются возможности, а акцентируется внимание на том, что происходит _разделение_ по ключам объекта, которое очень можно удобно применить. А именно: те имена переменных, у которых нет префикса `...`, работают как "экстракторы" для соответствующих ключей. А та переменная, где `...` присутствует, будут включать в себя _остальные_ ключи/значения.

Кстати, мне нравится твоя идея применения явных указаний `...`, `layout` и `rest`.

@gcor я пока поменял на вариант, на основе твоего. Он чуть более многословен, но мне кажется попонятнее и оставляет меньше места для двусмысленности:

> **Хорошо:** Оператор расширения (`...`) помогает извлечь часть пропсов (например, `layout`) каждый в отдельную переменную, а оставшиеся — поместить в общую переменную (скажем, `rest`).

@gcor если это недостаточно хорошо и настаиваешь на своём варианте, я готов принять

> Оператор расширения (...) помогает извлечь часть пропсов (layout) и оставшиеся поместить в переменную (rest).

Reviewer: Мне нравится твой вариант. 
Я не очень силён в запятых, но перед _каждый_ возможно нужна.

Author: @gcor я поднял вопрос на gramota.ru 😆 . Не знаю, нужна ли запятая.

---
title: "Предупреждение: неизвестный проп"
layout: single
permalink: warnings/unknown-prop.html
---

> Предупреждение о неизвестном пропе выдаётся, когда вы пытаетесь срендерить DOM-элемент с пропом, который React не может распознать как разрешённый DOM-атрибут или свойство. Вы должны убедиться, что DOM-элементы не получают по ошибке не относящиеся к ним пропсы.

Reviewer 3: ```suggestion
Предупреждение о неизвестном пропе выдаётся, когда вы пытаетесь срендерить DOM-элемент с пропом, который React не может распознать как разрешённый DOM-атрибут или свойство. Вам следует убедиться, что DOM-элементы не получают по ошибке не относящиеся к ним пропсы.
```

Есть несколько наиболее вероятных причин, из-за чего возникает это предупреждение:

1. Используете ли вы `{...this.props}` или `cloneElement(element, this.props)`? Может быть, ваш компонент передаёт собственные пропсы напрямую дочернему элементу (см. [Компоненты и пропсы](/docs/transferring-props.html)). Проверьте, что вы не перенаправляете случайно пропсы, предназначенные для родительского компонента, в дочерние компоненты.

> 2. Вы используете нестандартный DOM-атрибут на нативном DOM-узле. Например, с целью представить пользовательские (custom) данные. Вместо этого вам следует рассмотреть вариант с использованием data-* атрибутов, описанных [на MDN](https://developer.mozilla.org/ru/docs/Web/Guide/HTML/Using_data_attributes).

Reviewer 3: ```suggestion
2. Вы используете нестандартный DOM-атрибут на нативном DOM-узле. Например, с целью представить пользовательские данные. Вместо этого следует рассмотреть вариант с использованием data-атрибутов, описанных [на MDN](https://developer.mozilla.org/ru/docs/Web/Guide/HTML/Using_data_attributes).
```

Author: ![image](https://user-images.githubusercontent.com/5762340/53680105-1b70aa80-3c8b-11e9-894c-213768b09fc1.png)

Author: @lex111 data-* атрибуты звучит плохо, да?

Reviewer 3: Выглядит странно, поэтому [обычный](https://www.google.com/search?newwindow=1&biw=1853&bih=927&ei=H056XLz9DaWOrwTzrK6AAg&q=%22data-%D0%B0%D1%82%D1%80%D0%B8%D0%B1%D1%83%D1%82%D0%BE%D0%B2%22+javascript&oq=%22data-%D0%B0%D1%82%D1%80%D0%B8%D0%B1%D1%83%D1%82%D0%BE%D0%B2%22+javascript&gs_l=psy-ab.3..33i22i29i30j33i10.5200.9310..9448...1.0..0.200.1345.3j7j1......0....1..gws-wiz.......0i71j0i22i30j33i160.xpPgSwVOhRY) вариант лучше.

Author: 👍

> 3. React не узнаёт заданный вами атрибут. С большой вероятностью, это будет исправлено в будущих версиях React. На сегодняшний же день React вычищает неизвестные атрибуты, поэтому их указание не приводит к их рендерингу.

Reviewer 3: ```suggestion
3. React не распознаёт указанный атрибут. С большой вероятностью, это будет исправлено в будущих версиях React. На сегодняшний же день React удаляет неизвестные атрибуты, поэтому они не приведут к рендерингу.
```

Author: @lex111 ну, тогда уж

> 3. React не распознаёт указанный атрибут. С большой вероятностью, это будет исправлено в будущих версиях React. На сегодняшний же день React удаляет неизвестные атрибуты, поэтому они не БУДУТ ОТРЕНДЕРЕНЫ.

Страдательная форма в конце.

Reviewer 3: Обновил.

> 4. Вы используете React-компонент, имя которого начинается со строчной, а не заглавной буквы. React интерпретирует его как DOM-тег, потому что [JSX-трасформация в React использует соглашение о регистре в наименованиях для отличия определённых пользователем компонентов от DOM-тегов](/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized).

Reviewer 3: ```suggestion
4. Вы используете React-компонент, имя которого начинается со строчной, а не заглавной буквы. React интерпретирует его как DOM-тег, потому что [JSX в React использует как верхний, так и нижний регистры букв для записи элементов, чтобы отличать пользовательские компоненты от DOM-тегов](/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized).
```

Author: Могу я настоять на варианте, ближе к исходному?

> 4. Вы используете React-компонент, имя которого начинается со строчной, а не заглавной буквы. React интерпретирует его как DOM-тег, потому что [JSX в React опирается на соглашение о регистре первых символов в именах для отличия пользовательских компонентов от DOM-тегов](/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized).

Или разбить на два предложения, но покороче?

> 4. Вы используете React-компонент, имя которого начинается со строчной, а не заглавной буквы. React интерпретирует его как DOM-тег, потому что [JSX в React опирается на соглашение для отличия пользовательских компонентов от DOM-тегов](/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized). Если имя начинается с заглавной буквы, то оно представляет компонент; а если со строчной — DOM-тег.

Во-первых, упоминание о _соглашении_ тут важно я считаю, потому что это немного философский вопрос (configuration vs convention).
Во-вторых, lower case vs. upper case convention ну никак не ложится у меня в голове на _...использует как верхний, так и нижний регистры букв для записи элементов, чтобы..._

Если для тебя этот комментарий критичен, можно выслушать ещё чьё-то мнение. 🤷‍♂️

Reviewer 3: Просто хотел подчернуть _lower case vs. upper case convention_, т.е. нижний и верхний регистры, потому что в соглашении они оба используются. Тогда реши, пожалуйста, сам это, у меня нет больше вариантов.

Author: Хорошо, но я подожду, если кто-то ещё отреагирует на замечание.

Reviewer: Может, например, так:

Имя React-компонента начинается не с заглавной буквы, поэтому React распознаёт его, как DOM-тег. Это происходит на основе соглашения о верхнем и нижнем регистре в JSX, которое позволяет React различать пользовательские компоненты от DOM-тегов.

Author: @gcor @lex111 спасибо! Исправил.

---

> Чтобы исправить ситуацию, ваш композитный компонент должен "поглотить" все пропсы, которые предназначены именно ему, а не его дочерним компонентам. Пример:

Reviewer 3: ```suggestion
Чтобы исправить ситуацию, составные компоненты должны получить все пропсы, которые предназначены только ему, а не его дочерним компонентам. Пример:
```

> **Плохо:** Непредвиденный проп `layout` перенаправлен а тег `div`.

Reviewer 3: ```suggestion
**Плохо:** Непредвиденный проп `layout` перенаправлен в тег `div`.
```

```js
function MyDiv(props) {

>   if (props.layout === 'горизонтальный') {

Reviewer 3: ```suggestion
  if (props.layout === 'horizontal') {
```

>     // Плохо! Потому что мы уверены, что "layout" не проп, который <div> понимает.

Reviewer 3: ```suggestion
    // Плохо! Потому что мы уверены, что "layout" не проп, с которым <div> работает.
```

    return <div {...props} style={getHorizontalStyle()} />
  } else {

>     // Плохо! Потому что мы уверены, что "layout" не проп, который <div> понимает.

Reviewer 3: ```suggestion
    // Плохо! Потому что мы уверены, что "layout" не проп, с которым <div> работает.
```

    return <div {...props} style={getVerticalStyle()} />
  }
}
```

> **Хорошо:** Оператор `...` («spread operator») может пригодиться, чтобы «вытянуть» отдельные переменные из пропсов и положить оставшиеся пропсы в переменную.

Reviewer 3: ```suggestion
**Хорошо:** Оператор расширения (`...`) может пригодиться, чтобы «вытянуть» отдельные переменные из пропсов, а оставшиеся пропсы поместить в переменную.
```

```js
function MyDiv(props) {
  const { layout, ...rest } = props

>   if (layout === 'горизонтальный') {

Reviewer 3: ```suggestion
  if (layout === 'horizontal') {
```

    return <div {...rest} style={getHorizontalStyle()} />
  } else {
    return <div {...rest} style={getVerticalStyle()} />
  }
}
```

> **Хорошо:** Вы также можете присвоить пропсы новому объекту и удалить из него те ключи, которые вы используете в текущем компоненте. Этот объект можно безопасно передать дочерним компонентам. Будьте внимательны и не удаляйте пропсы из оригинального объекта `this.props`, потому что с ним следует обращаться как с иммутабельным.

Reviewer 3: ```suggestion
**Хорошо:** Вы также можете присвоить пропсы новому объекту и удалить из него те ключи, которые используете в текущем компоненте. Этот объект можно безопасно передать дочерним компонентам. Будьте внимательны и не удаляйте пропсы из оригинального объекта `this.props`, потому что с ним следует обращаться как с иммутабельным.
```

```js
function MyDiv(props) {

  const divProps = Object.assign({}, props);
  delete divProps.layout;

>   if (props.layout === 'горизонтальный') {

Reviewer 3: ```suggestion
  if (props.layout === 'horizontal') {
```

Author: Эмм, мы же переводим литералы в коде? Или я что-то упустил?

Reviewer 3: Да, но тут немного другой случай. Реально, кто-то будет использовать кириллицу для значения пропса, как тут? Ну ок.

Reviewer 3: Посмотрел у других, не переводят они (китайцы, японцы).

Author: @lex111 Понял ход мыслей. Принимаю твой коммит.

}
```

> **Хорошо:** Вы также можете присвоить пропсы новому объекту и удалить из него те ключи, которые используете в текущем компоненте. Этот объект можно безопасно передать дочерним компонентам. Будьте внимательны и не удаляйте пропсы из оригинального объекта `this.props`, потому что с ним следует обращаться как с иммутабельным.

Reviewer: Мне кажется, тут нет причины-следствия, поэтому _потому что_ не нужно.

```suggestion
**Хорошо:** Вы также можете присвоить пропсы новому объекту и удалить из него те ключи, которые используете в текущем компоненте. Этот объект можно безопасно передать дочерним компонентам. Будьте внимательны и не удаляйте пропсы из оригинального объекта `this.props`, обращайтесь с ним как с иммутабельным.
```
